### PR TITLE
Module for graphnav

### PR DESCRIPTION
**Pull request by heuristicus**
- [x] Potentially needs to integrate #28 #30 #31 #40, but hopefully hold those until this has been merged?
- [ ] Partial regression because some of the functions used try_claim: navigate_to, _navigate_to, and navigate_route

### CODE REVIEW

**Comment by sktometometo** [comment]
@heuristicus Should I wait to continue my pull requests until this PR is merged?

My PRs currently left are below.

- https://github.com/bdaiinstitute/spot_wrapper/pull/40
- https://github.com/bdaiinstitute/spot_wrapper/pull/31
- https://github.com/bdaiinstitute/spot_wrapper/pull/28

I am goint to create some pull requests about GraphNav recording after these, But it would be convenient for you and me that I wait to do so because my pull requests may cause conflicts.

**Comment by heuristicus** [comment]
Thanks for the consideration. I think I'd prefer if we could wait for my this PR to be merged, it would make it easier in terms of conflicts as you say. Perhaps @amessing-bdai could comment as well.

**Comment by sktometometo** [comment]
@amessing-bdai Do you have any idea or comment about this ?

**Comment by amessing-bdai** [comment]
I agree about holding the other graph nav PRs for this as the moving target would make it harder to test. I agree that as long as we can run this PR without any issues in testing that the `try_claim` regression can be handled in another PR. I don't think it will be that hard to fix as you just need a small class to hold the lease client and keepalive.

**Comment by kaiyu-zheng** [comment]
I'm a bit loaded but I will test this when I get a chance.

**Comment by heuristicus** [comment]
> Attempted to start spot_ros2 driver while on this branch for spot_wrapper, and got:

Sorry, I didn't have the opportunity to run most of this on our robot to catch this. Fixed in the latest commit.

**Comment by jbarry-bdai** [comment]
@kaiyu-zheng can you either test this or work with @smayorga-bdai to test this?  we're starting to get blocked on this so let's try and get it in.

**Comment by smayorga-bdai** [comment]
I think theres some regressions. I was able to run our test_graphnav_services script on our internal repo "main" but after checking out heuristicus/modular-graphnav I got the following errors when uploading the graph.
```
[spot_ros2-1] [INFO] [1690995381.627716334] [Chara.spot_wrapper]: Loaded graph has 111 waypoints and 118 edges
[spot_ros2-1] [ERROR] [1690995381.628465313] [Chara.spot_ros2]: Exception Error:'SpotWrapper' object has no attribute '_current_waypoint_snapshots'; 
[spot_ros2-1]  Traceback (most recent call last):
[spot_ros2-1]   File "/workspaces/bdai/ws/build/spot_driver/spot_driver/spot_ros2.py", line 1959, in handle_graph_nav_upload_graph
[spot_ros2-1]     self.spot_wrapper._upload_graph_and_snapshots(request.upload_filepath)
[spot_ros2-1]   File "/workspaces/bdai/ws/build/spot_wrapper/spot_wrapper/wrapper.py", line 1704, in _upload_graph_and_snapshots
[spot_ros2-1]     self._current_waypoint_snapshots[
[spot_ros2-1] AttributeError: 'SpotWrapper' object has no attribute '_current_waypoint_snapshots'
[spot_ros2-1] 
```

**Comment by jbarry-bdai** [comment]
> I think theres some regressions. I was able to run our test_graphnav_services script on our internal repo "main" but after checking out heuristicus/modular-graphnav I got the following errors when uploading the graph.
> 
> ```
> [spot_ros2-1] [INFO] [1690995381.627716334] [Chara.spot_wrapper]: Loaded graph has 111 waypoints and 118 edges
> [spot_ros2-1] [ERROR] [1690995381.628465313] [Chara.spot_ros2]: Exception Error:'SpotWrapper' object has no attribute '_current_waypoint_snapshots'; 
> [spot_ros2-1]  Traceback (most recent call last):
> [spot_ros2-1]   File ".../spot_driver/spot_ros2.py", line 1959, in handle_graph_nav_upload_graph
> [spot_ros2-1]     self.spot_wrapper._upload_graph_and_snapshots(request.upload_filepath)
> [spot_ros2-1]   File ".../spot_wrapper/wrapper.py", line 1704, in _upload_graph_and_snapshots
> [spot_ros2-1]     self._current_waypoint_snapshots[
> [spot_ros2-1] AttributeError: 'SpotWrapper' object has no attribute '_current_waypoint_snapshots'
> [spot_ros2-1] 
> ```

Oh yeah we need to update `spot_ros2` to work with this PR (@bhung-bdai )...  Can we not test it until that is done @kaiyu-zheng ?

**Comment by heuristicus** [comment]
The main change needed should be to call functions using graphnav with `spot_wrapper.spot_graph_nav` rather than just the wrapper, and there may be functions whose name changed, but I don't think so.

**Comment by bhung-bdai** [comment]
> > I think theres some regressions. I was able to run our test_graphnav_services script on our internal repo "main" but after checking out heuristicus/modular-graphnav I got the following errors when uploading the graph.
> > ```
> > [spot_ros2-1] [INFO] [1690995381.627716334] [Chara.spot_wrapper]: Loaded graph has 111 waypoints and 118 edges
> > [spot_ros2-1] [ERROR] [1690995381.628465313] [Chara.spot_ros2]: Exception Error:'SpotWrapper' object has no attribute '_current_waypoint_snapshots'; 
> > [spot_ros2-1]  Traceback (most recent call last):
> > [spot_ros2-1]   File ".../spot_driver/spot_ros2.py", line 1959, in handle_graph_nav_upload_graph
> > [spot_ros2-1]     self.spot_wrapper._upload_graph_and_snapshots(request.upload_filepath)
> > [spot_ros2-1]   File ".../spot_wrapper/wrapper.py", line 1704, in _upload_graph_and_snapshots
> > [spot_ros2-1]     self._current_waypoint_snapshots[
> > [spot_ros2-1] AttributeError: 'SpotWrapper' object has no attribute '_current_waypoint_snapshots'
> > [spot_ros2-1] 
> > ```
> 
> Oh yeah we need to update `spot_ros2` to work with this PR (@bhung-bdai )... Can we not test it until that is done @kaiyu-zheng ?

I'm admittedly confused to the status of this. When I talked to @amessing-bdai, he mentioned there were multiple PRs blocking other things throughout this whole process. That's mainly why I've been hestitant to test it; I just don't know exactly what is dependent on what, so I don't know what order I'm supposed to be evaluating changes.

**Comment by bhung-bdai** [comment]
I'll talk to Kaiyu about it. Andrew mentioned that he had a HIL test for
graphnav.

Brandon

**Comment by jbarry-bdai** [comment]
@bhung-bdai he does and the above is the output I believe from running that test (which @smayorga-bdai did with @kaiyu-zheng 's guidance so he can help test this out now!).  The problem is that it relies on spot ros2, which isn't compatible with this PR :(

**Comment by bhung-bdai** [comment]
I can get on this. If it was like the spot_images modular change, then it
should be fairly simple to change spot_ros2 to work with it.

Brandon

**Comment by bhung-bdai** [comment]
Ah, that explains a lot. Thanks! I apologize for the spam.

Brandon

**Comment by heuristicus** [comment]
Should be corrected now.

**Comment by bhung-bdai** [comment]
Localization HIL test works with the update `spot_ros2`. I would like to test `navigation` functions before I approve it though.

**Comment by bhung-bdai** [comment]
Tested our HIL test with `spot_ros2` and `spot_wrapper`, also tested to make sure the navigation and graph list + upload functions still work. Added a minor bug fix.

**Review by bhung-bdai — APPROVED** [review]
LGTM

File: spot_wrapper/tests/test_graph_nav_util.py
Comment (mpickett-bdai) [inline spot_wrapper/tests/test_graph_nav_util.py:155]: Is there a difference between this test and the test above? I'm not as familiar with this code so maybe I'm missing something.

**Review by mpickett-bdai — APPROVED** [review]
LGTM s

**Review by baxelrod-bdai — APPROVED** [review]

**Pull request merged**
